pkg/admission/prep_downscale.go

```diff
@@ -0,0 +1,185 @@
+	// Changes from/to nil scale are not downscales strictly speaking.
+	if oldReplicas == nil || newReplicas == nil {
+		return allowWarn(logger, "old/new replicas is nil, allowing the change")
+	}
+	// If it's not a downscale, just log debug.
+	if *oldReplicas < *newReplicas {
+		level.Debug(logger).Log("msg", "upscale allowed")
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+	if *oldReplicas == *newReplicas {
+		level.Debug(logger).Log("msg", "no replicas change, allowing")
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+
+	// Get the resource labels: for example, for a StatefulSet, it will be the labels of the StatefulSet itself,
+	// while for a Scale object, it will be the labels of the StatefulSet/Deployment/ReplicaSet that the Scale object belongs to.
+	var lbls map[string]string
+	switch o := oldObj.(type) {
+	case *appsv1.Deployment:
+		lbls = o.Labels
+	case *appsv1.StatefulSet:
+		lbls = o.Labels
+	case *appsv1.ReplicaSet:
+		lbls = o.Labels
+	case *autoscalingv1.Scale:
+		lbls, err = getResourceLabels(ctx, ar, api)
+		if err != nil {
+			return allowBecauseCannotGetResource(ar, logger, err)
+		}
+	default:
+		return allowWarn(logger, fmt.Sprintf("unsupported type %T, allowing the change", o))
+	}
+
+	// Since it's a downscale, check if the resource has the label that indicates it's ready to be prepared to be downscaled.
+	// Create a slice of endpoint addresses for pods to send HTTP post requests to and to fail if any don't return 200
+	if !*ar.Request.DryRun && lbls[PrepDownscaleLabelKey] == PrepDownscaleLabelValue {
+		diff := (*oldReplicas - *newReplicas)
+		eps := make([]string, diff)
+
+		// The DNS entry for a pod of a stateful set is
+		// ingester-zone-a-0.$(servicename).$(namespace).svc.cluster.local
+		// The service in this case is ingester-zone-a as well.
+		// https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id
+		for i := 0; i < int(diff); i++ {
+			eps[i] = fmt.Sprintf("%v-%v.%v.%v.svc.cluster.local:%s/ingester/%s",
+				ar.Request.Name,       // pod name
+				int(*oldReplicas)-i-1, // nr in statefulset
+				ar.Request.Name,       // svc name
+				ar.Request.Namespace,
+				lbls[PrepDownscalePortKey],
+				lbls[PrepDownscalePathKey],
+			)
+		}
+
+		g, _ := errgroup.WithContext(ctx)
+		for _, ep := range eps {
+			ep := ep // https://golang.org/doc/faq#closures_and_goroutines
+			g.Go(func() error {
+				resp, err := client.Post("http://"+ep, "application/json", nil)
+				if err != nil {
+					level.Error(logger).Log("msg", "error sending HTTP post request", "endpoint", ep, "err", err)
+					return err
+				}
+				if resp.StatusCode != 200 {
+					err := fmt.Errorf("HTTP post request returned non-200 status code")
+					level.Error(logger).Log("msg", err, "endpoint", ep, "status", resp.StatusCode)
+					return err
+				}
+				if resp.StatusCode == 200 {
+					// TODO(jordanrushing): set a label on the pod to indicate it's been prepared for shutdown? Or do we need to do that?
```

Comment:
I think it's a good idea. There was some questions about how to persist the fact that a pod has been asked to prepare for downscaling. E.g. if a pod prepares for downscaling, then crashes, restarts and doesn't realize it has been asked to prepare for downscaling, is shutdown but _doesn't_ do any of the special "prepare for downscale" logic.

The situation above requires the application to know it's running in Kubernetes and inspect annotations which isn't great but I'm not sure there's a good way around that.

Reply:
We can use the https://kubernetes.io/docs/concepts/workloads/pods/downward-api/ to expose it as an environment variable.

Reply:
I think it's interesting to set an _annotation_, not really a label.

>  E.g. if a pod prepares for downscaling, then crashes, restarts and doesn't realize it has been asked to prepare for downscaling, is shutdown but doesn't do any of the special "prepare for downscale" logic.

I described this scenario in the (internal) design doc, and yes, the ingester should somehow persist the fact that it's prepared for downscale.

> We can use the https://kubernetes.io/docs/concepts/workloads/pods/downward-api/ to expose it as an environment variable.

Unless there's something I'm missing, this would require an ingester restart, which we don't want to do here.

Reply:
> Unless there's something I'm missing, this would require an ingester restart, which we don't want to do here.

Looks like the downward API can also be used to [populate files](https://kubernetes.io/docs/concepts/storage/volumes/#downwardapi) and supports pod annotations. I'll look into this further and see if we can use it.

Reply:
Confirmed that this will work in our environment. Changes to pod annotations are reflected in the on-disk file.

```
~ # cat /etc/podinfo/annotations 
cni.projectcalico.org/containerID="<redacted>"
cni.projectcalico.org/podIP="<redacted>/32"
cni.projectcalico.org/podIPs="<redacted>/32"
kubernetes.io/config.seen="2023-04-10T16:20:28.230821451Z"
kubernetes.io/config.source="api"
```